dialer/recorder/recorder.go

```diff
@@ -0,0 +1,215 @@
+	if err != nil {
+		fmt.Println(err)
+		return nil, err
+	}
+
+	conn, err = dialerWithLocalAddr.DialContext(ctx, network, addr)
+	if err != nil {
+		return nil, err
+	}
+
+	return NewConnectionRecorder(path.Join(d.dir, fmt.Sprintf("%s-%d", addr, sourcePort)), conn)
+}
+
+func NewConnectionRecorder(fname string, conn net.Conn) (net.Conn, error) {
+	fd_writes, err := os.OpenFile(fname+"Writes", os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0666)
+	if err != nil {
+		return nil, err
+	}
+	fd_reads, err2 := os.OpenFile(fname+"Reads", os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0666)
+	if err2 != nil {
+		return nil, err2
+	}
+	return &ConnectionRecorder{fd_writes: fd_writes, fd_reads: fd_reads, orig: conn, write_record: &RecordWriter{new: true}, read_record: &RecordWriter{new: true}}, nil
+}
+
+type RecordWriter struct {
+	new       bool
+	to_record int
+	record    map[string]interface{}
+}
+
+type ConnectionRecorder struct {
+	fd_writes    *os.File
+	fd_reads     *os.File
+	orig         net.Conn
+	read_record  *RecordWriter
+	write_record *RecordWriter
+}
+
+func (c *ConnectionRecorder) Read(b []byte) (n int, err error) {
+	n, err = c.orig.Read(b)
+	if err != nil && err != io.EOF {
+		return n, err
+	}
+
+	// Only record if RESULT frame
+	if b[4] != 0x08 {
+		return n, err
+	}
+
+	if c.read_record.new {
+		stream_id := int(int16(b[2])<<8 | int16(b[3]))
+		c.read_record.record = make(map[string]interface{})
+		c.read_record.record["stream_id"] = stream_id
+		c.read_record.record["data"] = b[:n]
+
+		c.read_record.to_record = int(b[8])
+		if n < c.read_record.to_record {
+			c.read_record.new = false
+			c.read_record.to_record = c.read_record.to_record - n
+		} else {
+			// Write JSON record to file
+			jsonData, marshalErr := json.Marshal(c.read_record.record)
+			if marshalErr != nil {
+				return n, fmt.Errorf("failed to encode JSON record: %w", marshalErr)
+			}
+			_, writeErr := c.fd_reads.Write(append(jsonData, '\n'))
+			if writeErr != nil {
+				return n, fmt.Errorf("failed to record read: %w", writeErr)
+			}
+		}
+
+		return n, err
+	}
+
+	c.read_record.record["data"] = append(c.read_record.record["data"].([]byte), b[:n]...)
+
+	c.read_record.to_record = c.read_record.to_record - n
+	if c.read_record.to_record <= 0 {
+		c.read_record.new = true
+		// Write JSON record to file
+		jsonData, marshalErr := json.Marshal(c.read_record.record)
+		if marshalErr != nil {
+			return n, fmt.Errorf("failed to encode JSON record: %w", marshalErr)
+		}
+		_, writeErr := c.fd_reads.Write(append(jsonData, '\n'))
+		if writeErr != nil {
+			return n, fmt.Errorf("failed to record read: %w", writeErr)
+		}
+	}
+
+	return n, err
+}
+
+func (c *ConnectionRecorder) Write(b []byte) (n int, err error) {
+	n, err = c.orig.Write(b)
+
+	// Only record if EXECUTE frame
+	if b[4] != 0x0A {
+		return n, err
+	}
+
+	if c.write_record.new {
+		stream_id := int(b[2])<<8 | int(b[3])
+		b[2] = byte('0')
+		b[3] = byte('0')
+		writeHash := murmur.Murmur3H1(b[:9])
```

Comment:
I believe hash should be calculated in a different way for different frame types.
It is better to make a function of it `getFrameHash(frame []byte) int64` to be used here and on replay